src/onchain_reading.ts

```diff
@@ -0,0 +1,105 @@
+import {
+  decodeOrders,
+  decodeIndexedOrders,
+  Order,
+  IndexedOrder,
+} from "./encoding";
+import { BatchExchangeViewer } from "../build/types/BatchExchangeViewer";
+import { BatchExchange } from "../build/types/BatchExchange";
+import BN from "bn.js";
+
+/**
+ * Returns an iterator yielding an item for each page of order in the orderbook that is currently being collected.
+ * @param {BatchExchangeViewer} contract to query from
+ * @param {number} pageSize the number of items to fetch per page
+ * @param {number?} blockNumber the block number to execute the query at, defaults to "latest" if omitted
+ */
+export const getOpenOrdersPaginated = async function* (
+  contract: BatchExchangeViewer,
+  pageSize: number,
+  blockNumber: number | null,
+): AsyncGenerator<IndexedOrder<BN>[]> {
+  let nextPageUser = "0x0000000000000000000000000000000000000000";
+  let nextPageUserOffset = "0";
+  let hasNextPage = true;
+
+  if (blockNumber) {
+    contract = contract.clone();
+    contract.defaultBlock = blockNumber;
+  }
+
+  while (hasNextPage) {
+    const page = await contract.methods
+      .getOpenOrderBookPaginated([], nextPageUser, nextPageUserOffset, pageSize)
+      .call();
+    const elements = decodeIndexedOrders(page.elements);
+    yield elements;
+
+    //Update page info
+    hasNextPage = page.hasNextPage;
+    nextPageUser = page.nextPageUser;
+    nextPageUserOffset = page.nextPageUserOffset;
+  }
+};
+
+/**
+ * Returns open orders in the orderbook.
+ * @param {BatchExchangeViewer} contract to query from
+ * @param {number} pageSize the number of items to fetch per page
+ * @param {number?} blockNumber the block number to execute the query at, defaults to "latest" if omitted
+ */
+export const getOpenOrders = async (
```

Comment:
Why does this not use the `export function foo` syntax and instead uses `export const foo = ...`?

Reply:
I would (also?) prefer the `export function foo` syntax here.

Reply:
I, on the other hand, simply look forward to seeing what `export function foo` winds up looking like.

Reply:
Sure. There was no specific reason, I just applied the minimum number of keystrokes to get it to compile. Happy to remove the const.